lib/aws-manager.js

```diff
@@ -1247,7 +1125,9 @@ class AwsManager {
 
     let perRegionKills = {};
 
-    for (let region of this.ec2.regions) {
+    let killPromises = [];
+
+    for (let region of _.keys(this.ec2)) {
```

Comment:
We should convert this into `Promise.all(_.map(this.ec2` and avoid the `perRegionKills` data structure...

I've proposed something like this before, but here it is again. The `SmartFilter` a better name is welcome:
```js
/**
 * A [].filter-like method that takes a list and an object of conditions.
 *
 * SmartFilter([{...}, {...}, ...list of objects], {
 *   'a.property.path': value // a value that must be matched
 *   'property': val => ... // function that evaluates value at path...
 *   'sub.prop': [1, 2, 3]  // a list of allowed values, must be one of  
 * });
 *
 * Each property in the condition object is a condition. All conditions must
 * be satisfied. Each condition is a key, value pair of the condition object.
 * The key is an object path, the value is a value that must be matched, a
 * function that must return true, or a list which must contain the value.
 */
SmartFilter = (items, filter) => {
  return _.filter(items, item => {
    return _.every(filter, (val, key) => {
      if (val instanceof Array) {
        return _.includes(val, _.get(item, key));
      } else if (val instanceof Function) {
        return val(_.get(item, key));
      } else {
        return _.isEqual(_.get(item, key), val);
      }
    });
  });
};
```

```js
await Promise.all(_.map(this.ec2, async (ec2, region) => {
  let instanceIds = SmartFilter(this.__apiState.instances, {
    'WorkerType: name,
    'State.Name': states, // this is a one-of, see the implementation of SmartFilter
    'Region': region,
  }).map(_.property('InstanceId'));

  let requests = [];  
  .... // smilar hacks... probably using _.flatten to merge the two lists.. As you'll call SmartFilter twice

  await this.killCancel(region, instanceIds, requests);
}));
```

I know `SmartFilter(items, conditions)` might take a few seconds to understand.. And yes, the code isn't much shorter... In fact it's far more in-efficient because we iterate over all instances once for each region...

But all of that doesn't matter.. The code is much more declarative and functional. We declare how we want the list of instances filtered, rather than doing the actual filtering.

@jhford, please seriously consider this `SmartFilter` approach, it'll make so many things much clearer... I'll be happy to discuss it with you... I think it can be used many places in the provisioner code based, because we're operating on lists of objects, and we want to filter them..

(but manual filtering is boring, error prone, and **hard to read**, so let's try to be declarative).